generative-deep-learning/large-language-models/prompting-mistral7B.ipynb

```diff
@@ -0,0 +1,1594 @@
+    "    print(text, end=\"\", flush=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "79335c8f-ede7-4a65-813e-ec225822d5db",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "<div class=\"alert alert-block alert-info\"><b>The model did pretty well. Notice it started with generating a period (\".\"). That's because LLMs are trained to generate next most probably token, given an input. Let's see if we can make the answer more concise, and we will end our prompt with a period this time.</b></div>\n",
+    "\n",
+    "## Modifying the basic prompt\n",
+    "Below we will try giving the model more specific instructions.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "18fe60fe-bc65-4855-819e-a7ecb3071e4d",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      " Photosynthesis is the process by which plants, algae, and some bacteria convert sunlight, water, and carbon dioxide into glucose, oxygen, and other chemical compounds. This process occurs in the chloroplasts of plant cells and involves two stages: the light-dependent reactions, which use light energy to produce ATP and NADPH, and the light-independent reactions, also known as the Calvin cycle, which use these energy-rich molecules to fix carbon dioxide into glucose. Photosynthesis is a vital process for life on Earth, providing the primary source of food and oxygen for nearly all living organisms."
+     ]
+    }
+   ],
+   "source": [
+    "prompt = \"Explain photosythesis in three sentences.\"\n",
+    "\n",
+    "for text in llm(prompt, max_new_tokens=250, temperature=0.1, stream=True):\n",
+    "    print(text, end=\"\", flush=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a7071987-b179-4cd2-900c-28e02894c7d3",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "\n",
+    "<div class=\"alert alert-block alert-info\"><b>This response is more concise, our prompt worked quite well.</b></div>\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "64dc398e-65a6-4ed6-99bc-19a3f2fa0498",
+   "metadata": {},
+   "source": [
+    "# Fine-Tune with prompt\n",
+    "\n",
+    "When interact with large lanauge models, you can enhance the model performance by Weight fine-tuning, and instruction fine-tuning.\n",
+    "\n",
+    "- Weight fine-tuning involves taking a pre-trained language model and continuing to train it on a downstream task, tweaking the weights of the model through additional training on task-specific data. This allows the model to become specialized for a particular application. One technique of weight fine-tuning is Low-Rank Adaptation(LoRA). Read more about it: [A Brief Introduction to LoRA: Low-Rank Adaptation of Large Language Models](https://wandb.ai/sauravmaheshkar/LoRA/reports/A-Brief-Introduction-to-LoRA-Low-Rank-Adaptation-of-Large-Language-Models--Vmlldzo2MDAyOTU3)\n",
```

Comment:
I think the fine-tuning explanation of this section is wrong.

At first, instruction tuning does not mean `keeps the weights of the pre-trained model`. For example, [`InstructGPT`](https://openai.com/research/instruction-following), the former version of ChatGPT, uses supervised fine tuning (full parameter tuning) for instruction tuning. Secondly, [LoRA tuning](https://arxiv.org/abs/2106.09685) is not the way for `Weight fine-tuning` because it is the way to keep pre-trained model weights freeze. LoRA is common way for instruction tuning.

I recommend to refer the lecture [Finetuning Large Language Models](https://www.deeplearning.ai/short-courses/finetuning-large-language-models/). It will be useful to explain the difference between `fine tuning` and `instruction tuning`. Here is the capture from the lecture.
![image](https://github.com/aws/studio-lab-examples/assets/544269/9f0e6d6a-bc08-4e58-ad94-f61ad9c28196)